System tests: Recovery

serrss opened the pull request:
### Type of change
- System tests

### Description
System tests to check recovery when some deployments or stateful sets are deleted.
Case 1: Test deletes deployment TC and waits for recovering
Case 2: Test deletes deployment CC and waits for recovering
Case 3: Test deletes Kafka stateful set and waits for recovering
Case 4: Test deletes Zookeeper stateful set and waits for recovering

###Additional information
Test cases 1, 2-4 are failed now by issue #264 

### Checklist
- [ ] Write tests
- [ ] Make sure all tests pass

ppatierno commented:
Well thinking more, at least it shouldn't happen for the `cluster-controller` deletion. In the other cases, the running cluster-controller should execute the reconciliation re-creating topic controller deployment and statefulsets for other nodes.

ppatierno commented on systemtest/src/test/java/io/strimzi/systemtest/KafkaClusterTest.java:
```diff
@@ -253,4 +259,66 @@ public void testClusterWithCustomParameters() {
+        kubeClient.waitForResourceDeletion(DEPLOYMENT, clusterControllerName);
+
+        LOGGER.info("Waiting for recovery {}", clusterControllerName);
+        kubeClient.waitForDeployment(clusterControllerName);
```
As I described in the main comment, the other tests should work thanks to the running cluster controller doing the reconciliation but this test should not work; no-one will recreate the cluster controller if you delete its related Deployment.

scholzj replied:
Yeah, this is valid point - there is nobody to recover the cluster controller. This test doesn't make sense. The other tests do.

serrss replied:
Absolutely agree, there is no sense to have a test case for deleting CC. I'm going to remove this case

scholzj commented:
@ppatierno This is basically a (partial since it doesn't delete the config maps or services) system test for #264 which @tombentley is currently working on.

serrss commented:
@scholzj @tombentley Absolutely agree that we should have cases for deletion of the metrics CM and Services in this PR. As for improvement for reconciliation interval, I suggest creating another issue and do it a little bit later.

tombentley commented on common-test/src/main/java/io/strimzi/test/StrimziRunner.java:
```diff
@@ -238,7 +240,9 @@ protected void before() {
                     // delete cm
-                    kubeClient().deleteContent(yaml);
+                    KubernetesClient client = new DefaultKubernetesClient();
+                    client.configMaps().withName(cluster.name()).delete();
```
Why was this change necessary? I don't think it works in practice, because although the kubeClient() is correctly namespaced (i.e. according to the `@Namespace`), the DefaultKubernetesClient() isn't. We probably want to provide access to a `NamespacedKubernetesClient` from the `clusterResource`.

serrss replied:
@tombentley Good catch! I debugged this code and looks like it works now because the necessary namespace is a first in an array of two existing namespaces. So I fixed this bug by using config for `DefaultKubernetesClient` with namespase which can be obtained from Strimzi `KubeClient`.

tombentley commented:
It will now be necessary to rebase and retest this, but in theory these new tests should now pass.

serrss commented:
Sorry folks @asorokhtey @scholzj @tombentley @ppatierno , but I did some changes in this PR and it requires your approval again =(
1. I created an abstract class for system tests to avoid duplication of code and have the ability to sort the system tests by equivalence class
2. Moved tests that check the recovery in a separate class
3. Removed fabrik8 client from `StrimziRunner` because tests are stable after merging PR with resiliency

Additional information:
The build takes ~50 mins after adding system tests for recovering and it's a reason that Travis CI job was failed. @scholzj any idea what is a right solution for this case? Should I skip tests?

scholzj commented:
@serrss Is the 50 minutes with already with smaller reconciliation timeout?

We can either enable them and live with it. Or disable them here and run them somewhere downstream in Jenkins.

serrss commented:
@scholzj  thanks for answering. Yes, it takes 50 mins with reconciliation timeout = 30 secs. I'm going to reduce this value to 5 secs but looks like it can save at least ~ 10 mins

scholzj commented:
That sounds reasonable ... deploying the cluster for every single test would take a lot of time.

tombentley commented on systemtest/src/test/java/io/strimzi/systemtest/RecoveryClusterTest.java:
```diff
@@ -0,0 +1,159 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(RecoveryClusterTest.class);
+
+    @Test
+    public void testDeleteTopicControllerDeployment() {
```
We should try to name these tests better. We're not testing that we *can* delete the TC, we're testing that we can *recover* from TC deletion. So something like `testRecoveryFromTopicControllerDeletion()` would be clearer imho.

serrss replied:
Yeah, it sounds much more reasonable! Will rename class and methods.
